cmd/zscrub/zscrub.c

```diff
@@ -0,0 +1,242 @@
+                {
+                    char **tmp = realloc(searchdirs,
+                                         (nsearchdirs + 1) * sizeof (char *));
+                    if (tmp == NULL) {
+                        (void) fprintf(stderr, "Out of memory\n");
+                        exit(1);
+                    }
+                    searchdirs = tmp;
+                    searchdirs[nsearchdirs++] = optarg;
+                }
+                break;
+            case 'r':
+                // Restart a scrub
+                do_restart++;
+                break;
+            case 'v':
+                // Be chatty
+                verbose++;
+                break;
+            case '?':
+                (void) fprintf(stderr, "invalid option '%c'\n", optopt);
+                exit(1);
+            }
+        }
+    }
+
+    if (optind == argc) {
+        (void) fprintf(stderr, "Need pool name\n");
+        exit(1);
+    }
+    if (optind + 1 < argc) {
+        (void) fprintf(stderr, "WARNING: Discarding excess arguments\n");
+    }
+    poolname = argv[optind];
+
+    if (verbose && (searchdirs != NULL)) {
+        int sdix = 0;
+        fprintf(stderr, "Will search:\n");
+        for (sdix = 0; sdix < nsearchdirs; sdix++) {
+            fprintf(stderr, "\t%s\n", searchdirs[sdix]);
+        }
+    }
+
+    dprintf_setup(&argc, argv);
+    kernel_init(FREAD | FWRITE | FCREAT);
+    VERIFY(libzfs_handle = libzfs_init());
+
+    {
+      importargs_t ia;
+      memset(&ia, 0, sizeof(ia));
+
+      ia = (struct importargs) {
+        .poolname = poolname,
+        .path = searchdirs,
+        .paths = nsearchdirs,
+        .can_be_active = 0,
+        .unique = 1,
+        .scan = 1,
+      };
+
+      nvlist_t *zpi = zpool_search_import(libzfs_handle, &ia);
+      if (nvlist_empty(zpi)) {
+          fprintf(stderr, "Unable to comply: cannot find pool\n");
+          exit(1);
+      }
+
+      nvpair_t *elem = nvlist_next_nvpair(zpi, NULL);
+      nvlist_t *config;
+      VERIFY(nvpair_value_nvlist(elem, &config) == 0);
+
+      error = spa_import(ia.poolname, config, NULL, ZFS_IMPORT_NORMAL);
+      if ((error != 0) && (error != EEXIST)) {
+          fprintf(stderr, "Unable to comply: cannot import '%s': %s\n",
+                  ia.poolname, strerror(error));
+          exit(2);
+      }
+
+      VERIFY3U(spa_open(ia.poolname, &spa, FTAG), ==, 0);
+
+      if (verbose) {
+        nvlist_t *nvroot;
+
+        fprintf(stderr, "Found pool; vdev tree:\n");
+        VERIFY3U(nvlist_lookup_nvlist(config,
+                   ZPOOL_CONFIG_VDEV_TREE, &nvroot), ==, 0);
+        print_vdev(ia.poolname, nvroot, 0);
+      }
+    }
+
+    spa->spa_debug = B_TRUE;
+
+    if (do_restart) {
+        if (verbose) { fprintf(stderr, "Cancelling any existing scrub...\n"); }
+        dsl_scan_cancel(spa->spa_dsl_pool);
+    }
+
+    if (verbose) { fprintf(stderr, "Kicking off scrub...\n"); }
+    spa_scan(spa, POOL_SCAN_SCRUB);
+    do {
```

Comment:
Would it be beneficial to have a way for the user to trigger a pause in the scrub? Might be useful to allow the pause/resume code to get exercised as well.